hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/HashJoinP.java

```diff
@@ -77,7 +81,30 @@ public HashJoinP(
         this.lookupTables = prependNull(Collections.nCopies(keyFns.size(), null));
         this.tags = tags.isEmpty() ? emptyList() : prependNull(tags);
         this.mapToOutputBiFn = mapToOutputBiFn;
-        this.mapToOutputTriFn = mapToOutputTriFn;
+
+        if (tags.isEmpty()) {
+            if (keyFns.size() == 2) {
+                this.flatMapper = flatMapper(object -> traverseStream(
```

Comment:
This code looks too complex to me. One example: `Stream.of(object).flatMap(o -> fun(o).stream())` is a redundant way of stating `fun(o).stream()`.

Reply:
@mtopolnik Thanks, indeed this code should be simplified. I've done that in next commit :)